### PR TITLE
[Kernel] Skip out-of-window KV blocks in paged decode sliding-window path

### PR DESCRIPTION
**Pull request by hec-ovi**
## Purpose

`kernel_paged_attention_2d` in `chunked_prefill_paged_decode.py` iterates
KV blocks from 0 to `num_blocks` and applies the sliding window as a score
mask after K/V are already loaded:

```python
if SLIDING_WINDOW > 0:
    S = tl.where((context_len - seq_offset) < SLIDING_WINDOW, S, -10000)
```

Decode is bandwidth-bound, so windowed layers pay full-context memory
traffic per decoded token even though everything outside the window is
discarded. For a model like poolside/Laguna-S-2.1 (36 of 48 layers with a
512-token window) at 16K context, roughly 97 percent of the K/V reads in
those layers are wasted. The unified attention kernel already bounds its
tile loop to the window (`compute_tile_loop_bounds`); this brings the
paged decode kernel used by the ROCM_ATTN backend up to par.

## Change

Start the block loop at the window edge instead of 0. The existing score
mask still trims the partial first block, and the masked-out contributions
it removes underflow to exactly 0 in the softmax accumulation, so outputs
are unchanged.

## Measurements

gfx1151 (Strix Halo), bf16, GQA 4:1, head_dim 128, window 512, single
decode token, kernel-level timing over 100 iterations. Outputs checked
against a plain PyTorch paged-attention reference (max abs err 0.002 for
both stock and patched, bf16 rounding):

| seq_len | stock | patched | speedup |
|---------|-------|---------|---------|
| 512     | 117 us | 116 us | 1.0x |
| 2048    | 221 us | 114 us | 1.9x |
| 8192    | 661 us | 112 us | 5.9x |
| 16384   | 1520 us | 114 us | 13.4x |

The patched kernel's latency is independent of context length, which is
the behavior the sliding window is supposed to provide.

## Notes

- No behavior change for `SLIDING_WINDOW == 0` (start_block stays 0).
- ALiBi and sink paths are unaffected: sinks are accumulated before the
  loop, and ALiBi biases only apply to tokens inside the window.
- The 2D kernel only runs for single-token queries (`query_len == 1`
  guard at the top), so `context_len = seq_len - 1` is the query position
  and the window arithmetic matches the existing mask exactly.

### CODE REVIEW

**Comment by cadamcat** [comment]
We rediscovered this change independently about a month after this PR, and found this PR via the duplicate check in `AGENTS.md` before opening anything: our eleven lines are identical to yours once the identifier and the `tl.maximum` argument order are normalised. Rather than open a second PR, here is a second, non-overlapping body of evidence, in case it helps this move out of draft.

**End to end, two real models, gfx1100 (2× RX 7900 XT, TP=2), three processes per cell, seven depths.** Milliseconds per generated token at 32 768 context: `gemma-3-27b` (w4a16, window 1 024) 124.29 → 45.26, **2.75×**; `Muse-Glimmer-30B` (int4, window 2 048) 83.99 → 26.63, **3.15×**. Below each model's own window the change is worth exactly 1.00×, and the gain grows monotonically past it — the shape a masked-block skip must produce, so the curve doubles as a mechanism check. In a full campaign the patched `Muse-Glimmer-30B` decodes essentially flat from its window onward (37.4 tok/s at 32 K), and `gemma-3-27b` goes 8.05 → 22.05 tok/s at 32 K.

**Correctness.** Upstream's own `tests/kernels/attention/test_prefix_prefill.py`, version-matched to the installed tree, run whole before and after and compared per case from junit XML: 388 cases, zero changed outcome, zero appearing in only one run (the suite itself runs 2.6× faster patched). Plus 15 boundary cases — `first_block` straddling 0 and 1, `seq_len` at window ± 1, an unaligned window, a mixed batch — built the way `tests/kernels` builds inputs and compared across separate processes with `torch.equal`: **bit-identical**, which is the right bar here, since skipped blocks contribute `exp(−10000 − m) = 0` and the skip is an identity rather than an approximation.

**Kernel level**, same profiler settings both sides: `kernel_paged_attention_2d` goes 1.589 ms → 154.796 µs per call at 32 K — 10.3×, short of the 16× the block count implies, because per-call overhead does not shrink with the loop.

Scope note: the models that pay this cost are sliding-window models whose decode lands on `ROCM_ATTN`'s Triton kernel — currently the sinks-free ones like the two above. `gemma-4` is forced onto `TRITON_ATTN` by its heterogeneous head dims and never pays it, and models with sinks also select `TRITON_ATTN` by default, whose decode loop is already window-bounded. Full write-up, raw data and per-depth tables: [sliding-window-block-skip.md](https://github.com/cadamcat/dual-radeon-vllm/blob/main/docs/sliding-window-block-skip.md).
